extensions-core/hdfs-storage/src/main/java/io/druid/storage/hdfs/HdfsDataSegmentKiller.java

```diff
@@ -48,42 +48,80 @@ public HdfsDataSegmentKiller(final Configuration config, final HdfsDataSegmentPu
     this.storageDirectory = new Path(pusherConfig.getStorageDirectory());
   }
 
+  private static Path getPath(DataSegment segment)
+  {
+    return new Path(String.valueOf(segment.getLoadSpec().get(PATH_KEY)));
+  }
+
+  // returns path up to data source for a segment
+  private static Path getDataSourcePath(DataSegment segment)
+  {
+    final String dataSource = segment.getDataSource();
+    final String segmentPath = getPath(segment).toUri().toString();
+
+    StringBuilder builder = new StringBuilder(segmentPath.split(dataSource)[0]);
```

Comment:
This isn't robust, what if the segmentPath is `/data/druid/segments/druid/20111001T000000.000Z_20111002T000000.000Z/20111001T000000.000Z/index.zip` because the dataSource name is "druid"? Then this method will return `/data/druid` which isn't right.

One alternative is to have two different deleteEmptyParents logics, one for things named like "index.zip" and one for things like "0_index.zip", using the old safeNonRecursiveDelete method.